forge/examples/torch/train.py

```diff
@@ -0,0 +1,126 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+
+
+from __future__ import print_function
+from os import path as osp
+
+import torch
+import torch.nn.functional as F
+import torch.optim as optim
+
+import forge
+from forge import flags
+import forge.experiment_tools as fet
+
+
+# Job config
+flags.DEFINE_string('data_config', 'configs/mnist_data.py',
+                    'Path to a data config file.')
+flags.DEFINE_string('model_config', 'configs/mnist_mlp.py',
+                    'Path to a model config file.')
+flags.DEFINE_string('results_dir', 'checkpoints',
+                    'Top directory for all experimental results.')
+flags.DEFINE_string('run_name', 'mnist',
+                    'Name of this job and name of results folder.')
+flags.DEFINE_boolean('resume', False, 'Tries to resume a job if True.')
+
+# Logging config
+flags.DEFINE_integer('report_loss_every', 100,
+                     'Number of iterations between reporting minibatch loss.')
+flags.DEFINE_integer('train_epochs', 20, 'Maximum number of training epochs.')
+
+# Experiment config
+flags.DEFINE_integer('batch_size', 32, 'Mini-batch size.')
+flags.DEFINE_float('learning_rate', 1e-5, 'SGD learning rate.')
+
+# Parse flags
+config = forge.config()
+
+# Prepare enviornment
+logdir = osp.join(config.results_dir, config.run_name)
+logdir, resume_checkpoint = fet.init_checkpoint(
+    logdir, config.data_config, config.model_config, config.resume)
+checkpoint_name = osp.join(logdir, 'model.ckpt')
+
+# Load data
+train_loader = fet.load(config.data_config, config)
+# Load model
+model = fet.load(config.model_config, config)
+
+# Print flags
+fet.print_flags()
+# Print model info
+print(model)
+
+# Setup optimizer
+optimizer = optim.RMSprop(model.parameters(),
+                          lr=config.learning_rate,
+                          momentum=0.9)
+
+# Try to restore model and optimizer from checkpoint
+if resume_checkpoint is not None:
+    print("Restoring checkpoint from '{}'".format(resume_checkpoint))
+    checkpoint = torch.load(resume_checkpoint)
+    # Restore model
+    model.load_state_dict(checkpoint['model_state_dict'])
+    # Restore optimizer
+    optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
+    # Update starting epoch
+    start_epoch = checkpoint['epoch'] + 1
+else:
+    start_epoch = 1
+print("Starting training at iter = {}".format(start_epoch))
+
+# Training
+for epoch in range(start_epoch, config.train_epochs + 1):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        optimizer.zero_grad()
+        output = model(data.view(-1, 784))
+        loss = F.cross_entropy(output, target)
```

Comment:
What's `F`?

Reply:
This is from "import torch.nn.functional as F"

torch.nn defines (neural network) modules as classes and torch.nn.functional is the functional view. I typically use torch.nn for modules that have parameters (e.g. nn.Linear, nn.Conv2d) and torch.nn.functional for parameter-free "layers" (e.g. F.relu, F.cross_entropy)

Reply:
I'd suggest using at least 3-letter names, as per python style guide.

Reply:
This is quite a standard naming convention in the PyTorch community